rfcs/0038-unified-device-target-and-memory-scope-planning.md

```diff
@@ -0,0 +1,244 @@
+   a `TargetMap` from `DLDeviceType` to `Target`.
+4. Each call to a primitive operator for a particular `DLDeviceType` signals we need to compile
+   ('lower') that primitive for that device. The `Target` to use for that compilation is found from
+   the `TargetMap` by the `LowerTEPass`.
+
+For the BYOC flow things are quite different:
+1. Operators may be annotated with an `FTVMAnnotateTarget` function for a particular
+   `target.<name>`. Here `<name>` serves only to distinguish possible BYOC toolchain names and is
+   currently not connected to the `Target` machinery in any way. The function should return true if
+   the given expression could be compiled for toolchain `<name>`. (However there are currently no
+   examples of this annotation in-tree.)
+2. The `MergeComposite` pass can be used to assign a `"Composite"` attribute to Relay functions
+   which have been hoisted out of a larger expression based on a fusion pattern. The attribute can
+   have any value of the form `"some.arbitrary.prefix.<name>"`. Again, this indicates the function
+   could be compiled for toolchain `<name>`. (The EthosU compilation flow illustrates this approach
+   in-tree.)
+3. The `AnnotateTarget` pass looks for the annotations from (1) and (2) to decide the unique
+   toolchain name for every Relay sub-expression which should go via a BYOC path. The transitions in
+   to and out of those sub-expressions are marked with `compiler_begin` and `compiler_end`
+   annotations.
+4. The `PartitionGraph` pass hoists sub-expressions delimited by `compiler_begin` and `compiler_end`
+   annotations into new top-level `Function`s with a `"Compiler"` attribute bound to the toolchain
+   `<name>`.
+5. The rest of the compilation flow treats `"Compiler"` annotated functions specially.
+
+We have 6 problems:
+1. TVM is being targeted to environments with multiple CPUs (eg Arm 'Big.LITTLE') and multiple
+   tensor-friendly devices (eg a GPU as well as an accelerator such as Arm 'Ethos-U'). This means a
+   `DLDeviceType` no longer uniquely determines a `Target`.
+2. Though TVM's `Device` abstraction (an alias for `dlpack`'s `DLDevice`) is a pair of a
+   `DLDeviceType` and an arbitrary 'device id', TVM does not consistently plumb the device id
+   through annotations, passes and operators.  Thus currently we cannot use 'device id' to
+   distinguish, eg, two CPUs in the same system.
+3. Upcoming work requires us to distinguish and propagate memory scopes for data at the Relay
+   level. (See also [RFC #9](https://github.com/apache/tvm-rfcs/blob/main/rfcs/0009_Unified_Static_Memory_Planning.md)
+   which has a similar need for memory scope propagation at the TIR level). This is an identical
+   problem to propagating devices, and it seems most natural to simply combine targets, devices and
+   memory scopes into a single 'target of device planing' rather than implementing a whole new pass.
```

Comment:
i agree, but just clarifying we aren't using a single identifier to describe both the device and the memory scope?